common/Obelisk/OAuth/Route.hs

```diff
@@ -0,0 +1,122 @@
+-}
+module Obelisk.OAuth.Route where
+
+import           Prelude                       hiding ((.))
+
+import           Control.Categorical.Bifunctor (first)
+import           Control.Category              ((.))
+import           Control.Category.Monoidal     (coidl)
+import           Control.Monad.Error.Class     (MonadError)
+import           Data.Aeson                    (FromJSON, ToJSON)
+import           Data.Map                      (Map)
+import qualified Data.Map                      as Map
+import           Data.Text                     (Text)
+import           GHC.Generics                  (Generic)
+
+import           Obelisk.OAuth.State           (OAuthState, oAuthStateAsText,
+                                                unsafeMkOAuthState)
+import           Obelisk.Route
+import           Obelisk.Route.TH
+
+-- | Id of an OAuth client.
+--
+--   As specified <https://tools.ietf.org/html/rfc6749#section-2.2 here>.
+newtype OAuthClientId = OAuthClientId { unOAuthClientId :: Text }
+  deriving (Show, Read, Eq, Ord, Generic, ToJSON, FromJSON)
+
+
+-- | Code that can be used for retrieving the actual access token, by providing
+--   the client secret.
+--
+--   See `AuthorizationResponseType` for more details.
+--
+--   Note: Depending on the value of `_authorizationRequest_responseType ` this
+--   might actually be a proper OAuth token and not a code.
+newtype OAuthCode = OAuthCode { unOAuthCode :: Text }
+  deriving (Show, Read, Eq, Ord, Generic, ToJSON, FromJSON)
+
+
+-- | Actual access token as delivered by the backend.
+--
+--
+newtype OAuthToken = OAuthToken { unOAuthToken :: Text }
+  deriving (Show, Read, Eq, Ord, Generic, ToJSON, FromJSON)
+
+
+-- | Parameters that the authorization server is expected to provide when
+-- granting an authorization code request. See section
+-- <https://tools.ietf.org/html/rfc6749#section-4.1.2 4.1.2> of the
+-- specification.
+data RedirectUriParams = RedirectUriParams
+  { _redirectUriParams_code  :: OAuthCode
+  , _redirectUriParams_state :: OAuthState
+  }
+  deriving (Show, Read, Eq, Ord, Generic)
+
+
+-- | The OAuth routes necessary for authorization code grants. This should be
+--   made a sub-route of the client application both in frontend routes _and_
+--   backend routes.
+--
+-- The frontend route will be used for the redirect coming from the
+-- authorization server. The backend route is needed, so the frontend can
+-- request the backend to retrieve the actual access token given the
+-- transmitted authorization code. This extra step is needed as only the
+-- backend can and should know the client secret needed for retrieving the
+-- actual access token.
+data OAuthRoute :: * -> * where
+  -- This is a `Maybe` because we don't want to crash the application just
+  -- because of missing parameters. The `Nothing` case should be handled
+  -- gracefully and the user should be informed that the handshake failed.
+  OAuthRoute_TransmitCode :: OAuthRoute (Maybe RedirectUriParams)
```

Comment:
We should add a warning to the docs here that specifies that we *rely* on the fact that the encoder for this won't have a query string when the value is Nothing, and that the RedirectUriParams will be represented as a query string when it's Just.  There are probably some other requirements, as well.